### PR TITLE
KAFKA-12175 Migrate generator module to junit5

### PR DESCRIPTION
**Pull request by bertber**
Migrate generator module to junit5.

### CODE REVIEW

**Comment by bertber** [comment]
@chia7712

File: generator/src/test/java/org/apache/kafka/message/CodeBufferTest.java [hunk 1]
```diff
@@ -67,7 +68,7 @@ public void testIndentMustBeNonNegative() {
         try {
             buffer.decrementIndent();
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/CodeBufferTest.java]: Could you rewrite it by ```assertThrows```

**Review by chia7712 — COMMENTED** [review]
@bertber You have to update the build.gradle. 

1) remove generator from https://github.com/apache/kafka/blob/trunk/build.gradle#L245

2) remove ```testCompile libs.junitJupiterApi``` and ```testCompile libs.junitVintageEngine``` from https://github.com/apache/kafka/blob/trunk/build.gradle#L1082

3) add ```testCompile libs.junitJupiter```

**Comment by bertber** [comment]
@chia7712 Could you take a look? Thank you.

File: generator/src/test/java/org/apache/kafka/message/MessageGeneratorTest.java [hunk 2]
```diff
@@ -17,30 +17,28 @@
...
+        assertEquals(MessageGenerator.capitalizeFirst(""), "");
+        assertEquals(MessageGenerator.capitalizeFirst("abC"), "AbC");
     }
 
     @Test
     public void testLowerCaseFirst() throws Exception {
-        assertEquals("", MessageGenerator.lowerCaseFirst(""));
-        assertEquals("fORTRAN", MessageGenerator.lowerCaseFirst("FORTRAN"));
-        assertEquals("java", MessageGenerator.lowerCaseFirst("java"));
+        assertEquals(MessageGenerator.lowerCaseFirst(""), "");
```
Comment (g1geordie) [inline generator/src/test/java/org/apache/kafka/message/MessageGeneratorTest.java]: @bertber left a little comment .
we don't need to change the position of args.
assertEquals  only change when it contain message arg.

File: generator/src/test/java/org/apache/kafka/message/EntityTypeTest.java [hunk 3]
```diff
@@ -17,15 +17,13 @@
...
 package org.apache.kafka.message;
 
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.Timeout;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
 
-import static org.junit.Assert.fail;
+import static org.junit.jupiter.api.Assertions.fail;
```
Comment (g1geordie) [inline generator/src/test/java/org/apache/kafka/message/EntityTypeTest.java]: same as @chia7712  said .
if you can use assertThrows to replace  expectException . awesome

File: generator/src/test/java/org/apache/kafka/message/CodeBufferTest.java (hunk 1 above)
Comment (bertber) [reply generator/src/test/java/org/apache/kafka/message/CodeBufferTest.java]: I tried to rewrite it by assertThrows.

File: generator/src/test/java/org/apache/kafka/message/MessageGeneratorTest.java (hunk 2 above)
Comment (bertber) [reply generator/src/test/java/org/apache/kafka/message/MessageGeneratorTest.java]: I reverted previous change.

File: generator/src/test/java/org/apache/kafka/message/EntityTypeTest.java (hunk 3 above)
Comment (bertber) [reply generator/src/test/java/org/apache/kafka/message/EntityTypeTest.java]: I tried to use assertThrows to replace expectException.

**Review by chia7712 — COMMENTED** [review]
@bertber Could you merge the trunk to your PR?

File: build.gradle [hunk 4]
```diff
@@ -1079,8 +1079,7 @@ project(':generator') {
     compile libs.jacksonDatabind
     compile libs.jacksonJDK8Datatypes
     compile libs.jacksonJaxrsJsonProvider
-    testCompile libs.junitJupiterApi
-    testCompile libs.junitVintageEngine
+    testCompile libs.junitJupiter
   }
 
   integrationTest {
```
Comment (ijuma) [inline build.gradle]: You can now remove this.

**Review by ijuma — COMMENTED** [review]
You can remove the `if` in the code below:

```groovy
     if (it.project.name != 'generator') {
        useJUnit {
          excludeCategories 'org.apache.kafka.test.IntegrationTest'
        }
      }
```

File: build.gradle (hunk 4 above)
Comment (ijuma) [reply build.gradle]: I pushed this change.

**Comment by ijuma** [comment]
I pushed a change with both of my suggestions above.

**Comment by ijuma** [comment]
@chia7712 will leave it to you to do the review and merge.

**Comment by ijuma** [comment]
@chia7712 Also, be aware that Jenkins merges trunk before running the PR tests. So, if the current build looks good, no manual merge is required for that reason (since there are no conflicts).

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 5]
```diff
@@ -92,16 +90,16 @@ public void testContains() {
 
     @Test
     public void testSubtract() {
-        assertEquals(Versions.NONE,
-            Versions.NONE.subtract(Versions.NONE));
+        assertEquals(Versions.NONE.subtract(Versions.NONE), 
+                     Versions.NONE);
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: Could you make them in single line?

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 6]
```diff
@@ -92,16 +90,16 @@ public void testContains() {
...
         assertEquals(newVersions(0, 0),
             newVersions(0, 0).subtract(Versions.NONE));
         assertEquals(newVersions(1, 1),
             newVersions(1, 2).subtract(newVersions(2, 2)));
         assertEquals(newVersions(2, 2),
             newVersions(1, 2).subtract(newVersions(1, 1)));
-        assertEquals(null,
-            newVersions(0, Short.MAX_VALUE).subtract(newVersions(1, 100)));
+        assertEquals(newVersions(0, Short.MAX_VALUE).subtract(newVersions(1, 100)),
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: How about replacing it by ```assertNull```?

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 7]
```diff
@@ -69,11 +67,11 @@ public void testIntersections() {
         assertEquals(newVersions(3, 3),
             newVersions(0, Short.MAX_VALUE).intersect(
                 newVersions(3, 3)));
-        assertEquals(Versions.NONE,
-            newVersions(9, Short.MAX_VALUE).intersect(
-                newVersions(2, 8)));
-        assertEquals(Versions.NONE,
-            Versions.NONE.intersect(Versions.NONE));
+        assertEquals(newVersions(9, Short.MAX_VALUE).intersect(
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: This change is unnecessary.

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 8]
```diff
@@ -69,11 +67,11 @@ public void testIntersections() {
...
+                newVersions(2, 8)), 
+                Versions.NONE);
+        assertEquals(Versions.NONE.intersect(Versions.NONE),
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: ditto

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 9]
```diff
@@ -57,8 +55,8 @@ public void testRoundTrips() {
     }
 
     private void testRoundTrip(Versions versions, String string) {
-        assertEquals(string, versions.toString());
-        assertEquals(versions, Versions.parse(versions.toString(), null));
+        assertEquals(versions.toString(), string);
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: the fist argument is still "expected value" so we don't need to switch the args. There are many similar occurrence.

File: generator/src/test/java/org/apache/kafka/message/VersionConditionalTest.java [hunk 10]
```diff
@@ -37,7 +35,7 @@ static void assertEquals(CodeBuffer buffer, String... lines) throws Exception {
         for (String line : lines) {
             expectedStringBuilder.append(String.format(line));
         }
-        Assert.assertEquals(expectedStringBuilder.toString(), stringWriter.toString());
+        Assertions.assertEquals(stringWriter.toString(), expectedStringBuilder.toString());
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionConditionalTest.java]: We don't use the prefix ```Assertions``` now.

Comment (bertber) [reply generator/src/test/java/org/apache/kafka/message/VersionConditionalTest.java]: I tried to remove the prefix Assertions.

File: generator/src/test/java/org/apache/kafka/message/VersionConditionalTest.java [hunk 11]
```diff
@@ -17,27 +17,25 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
+@Timeout(120)
 public class VersionConditionalTest {
-    @Rule
-    final public Timeout globalTimeout = Timeout.millis(120000);
 
-    static void assertEquals(CodeBuffer buffer, String... lines) throws Exception {
+    static void claimEquals(CodeBuffer buffer, String... lines) throws Exception {
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionConditionalTest.java:31]: the build fails due to this renaming.
```
/home/chia7712/kafka/generator/src/test/java/org/apache/kafka/message/IsNullConditionalTest.java:37: error: cannot find symbol
        VersionConditionalTest.assertEquals(buffer,
                              ^
  symbol:   method assertEquals(CodeBuffer,String,String,String)
  location: class VersionConditionalTest
/home/chia7712/kafka/generator/src/test/java/org/apache/kafka/message/IsNullConditionalTest.java:57: error: cannot find symbol
        VersionConditionalTest.assertEquals(buffer,
                              ^
  symbol:   method assertEquals(CodeBuffer,String,String,String,String,String)
  location: class VersionConditionalTest
/home/chia7712/kafka/generator/src/test/java/org/apache/kafka/message/IsNullConditionalTest.java:76: error: cannot find symbol
        VersionConditionalTest.assertEquals(buffer,
                              ^
  symbol:   method assertEquals(CodeBuffer,String,String,String)
  location: class VersionConditionalTest
/home/chia7712/kafka/generator/src/test/java/org/apache/kafka/message/IsNullConditionalTest.java:96: error: cannot find symbol
        VersionConditionalTest.assertEquals(buffer,
                              ^
  symbol:   method assertEquals(CodeBuffer,String)
  location: class VersionConditionalTest
/home/chia7712/kafka/generator/src/test/java/org/apache/kafka/message/IsNullConditionalTest.java:115: error: cannot find symbol
        VersionConditionalTest.assertEquals(buffer,
                              ^
  symbol:   method assertEquals(CodeBuffer,String,String,String)
  location: class VersionConditionalTest
5 errors

```

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 12]
```diff
@@ -41,10 +40,10 @@ private static Versions newVersions(int lower, int higher) {
 
     @Test
     public void testVersionsParse() {
-        assertEquals(Versions.NONE, Versions.parse(null, Versions.NONE));
-        assertEquals(Versions.ALL, Versions.parse(" ", Versions.ALL));
-        assertEquals(Versions.ALL, Versions.parse("", Versions.ALL));
-        assertEquals(newVersions(4, 5), Versions.parse(" 4-5 ", null));
+        assertEquals(Versions.parse(null, Versions.NONE), Versions.NONE);
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: please revert those changes as first argument should be "expected" value.

File: generator/src/test/java/org/apache/kafka/message/VersionsTest.java [hunk 13]
```diff
@@ -63,17 +62,13 @@ private void testRoundTrip(Versions versions, String string) {
...
     public void testIntersections() {
-        assertEquals(newVersions(2, 3),
-            newVersions(1, 3).intersect(
-                newVersions(2, 4)));
-        assertEquals(newVersions(3, 3),
-            newVersions(0, Short.MAX_VALUE).intersect(
-                newVersions(3, 3)));
-        assertEquals(Versions.NONE,
-            newVersions(9, Short.MAX_VALUE).intersect(
+        assertEquals(newVersions(1, 3).intersect(
```
Comment (chia7712) [inline generator/src/test/java/org/apache/kafka/message/VersionsTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@bertber Thanks for your patch. Please take a look at following minor comments.

**Review by chia7712 — APPROVED** [review]
@bertber thanks for this nice patch. +1

**Comment by chia7712** [comment]
```
org.apache.kafka.streams.integration.AdjustStreamThreadCountTest.shouldAddAndRemoveStreamThreadsWhileKeepingNamesCorrect
```

fixed flaky

**Comment by chia7712** [comment]
> be aware that Jenkins merges trunk before running the PR tests

@ijuma  it seems that jenkins does not merge trunk before running QA.

**Comment by ijuma** [comment]
@chia7712 What have you seen that implies that? This was certainly the case before, not sure if something changed at some point (I know because I've seen compiler errors that would not be possible without the merge).

**Comment by chia7712** [comment]
> What have you seen that implies that? 

The following error implies that recent fixes are not included.

```
java.lang.AssertionError: Expected all streams instances in [org.apache.kafka.streams.KafkaStreams@2d6a0fff] to be REBALANCING within 30000 ms, but the following were not: {org.apache.kafka.streams.KafkaStreams@2d6a0fff=RUNNING}
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:26)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.lambda$waitForApplicationState$12(IntegrationTestUtils.java:936)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:350)
	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:318)
	at org.apache.kafka.streams.integration.utils.IntegrationTestUtils.waitForApplicationState(IntegrationTestUtils.java:919)
	at org.apache.kafka.streams.integration.AdjustStreamThreadCountTest.shouldAddAndRemoveStreamThreadsWhileKeepingNamesCorrect(AdjustStreamThreadCountTest.java:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-9926/8/testReport/junit/org.apache.kafka.streams.integration/AdjustStreamThreadCountTest/?cloudbees-analytics-link=scm-reporting%2Ftests%2Ffailed

**Comment by ijuma** [comment]
I investigated a bit more after Jason noticed a similar issue regarding the target branch not being merged. I think this regressed when we moved to `Jenkinsfile` based PR builds. I tried to fix it:

image

**Pull request merged**